Encodings.md

```diff
@@ -155,39 +155,40 @@ Supported Types: INT32, INT64
+
 The header is defined as follows:
 ```
 <block size in values> <number of miniblocks in a block> <total value count> <first value>
 ```
- * the block size is a multiple of 128 stored as VLQ int
- * the miniblock count per block is a diviser of the block size stored as VLQ int the number of values in the miniblock is a multiple of 32.
- * the total value count is stored as a VLQ int
- * the first value is stored as a zigzag VLQ int
+ * the block size is a multiple of 128; it is stored as a ULEB128 int
+ * the miniblock count per block is a divisor of the block size such that their quotient, the number of values in a miniblock, is a multiple of 32; it is stored as a ULEB128 int
+ * the total value count is stored as a ULEB128 int
+ * the first value is stored as a zigzag ULEB128 int
 
 Each block contains
 ```
 <min delta> <list of bitwidths of miniblocks> <miniblocks>
 ```
- * the min delta is a VLQ int (we compute a minimum as we need positive integers for bit packing)
+ * the min delta is a zigzag ULEB128 int (we compute a minimum as we need positive integers for bit packing)
  * the bitwidth of each block is stored as a byte
  * each miniblock is a list of bit packed ints according to the bit width stored at the begining of the block
 
-Having multiple blocks allows us to escape values and restart from a new base value.
+To encode a block, we will:
 
-To encode each delta block, we will:
+1. Compute the differences between consecutive elements. For the first element in the block, use the last element in the previous block or, in the case of the first block, use the first value of the whole sequence, stored in the header.
 
-1. Compute the deltas
+2. Compute the frame of reference (the minimum of the deltas in the block). Subtract this min delta from all deltas in the block. This guarantees that all values are non-negative.
 
-2. Encode the first value as zigzag VLQ int
+3. Encode the frame of reference (min delta) as a zigzag ULEB128 int followed by the bit widths of the miniblocks and the delta values (minus the min delta) bit packed per miniblock.
 
-3. For each block, compute the frame of reference(minimum of the deltas) for the deltas. This guarantees
-all deltas are positive.
+Having multiple blocks allows us to adapt to changes in the data by changing the frame of reference (the min delta) which can result in smaller values after the subtraction which, again, means we can store them with a lower bit width.
 
-4. encode the frame of reference delta as VLQ int followed by the delta values (minus the minimum) encoded as bit packed per miniblock.
+If there are not enough values to fill the last miniblock, we pad the miniblock so that its length is always the number of values in a full miniblock multiplied by the bit width. The values of the padding bits are unspecified.
```

Comment:
"The values of the padding bits are unspecified." They used to be specified as 0-s, was there a writer that padded using noise? I don't see any reason for this change mentioned in the JIRA.

Reply:
Yes, if I am not mistaken, parquet-mr does this.

In `flushBlockBuffer`, it uses all of the `bitWidths` array [here](https://github.com/apache/parquet-mr/blob/dc61e510126aaa1a95a46fe39bf1529f394147e9/parquet-column/src/main/java/org/apache/parquet/column/values/delta/DeltaBinaryPackingValuesWriterForInteger.java#L114) and the `deltaBlockBuffer` array [here](https://github.com/apache/parquet-mr/blob/dc61e510126aaa1a95a46fe39bf1529f394147e9/parquet-column/src/main/java/org/apache/parquet/column/values/delta/DeltaBinaryPackingValuesWriterForInteger.java#L124) even if they were not completely overwritten since the last block, and it does not set them to zero afterwards.

Reply:
I wonder whether we should say instead something like this: Writers should always pad with zeroes, but readers must accept paddings consisting of arbitrary bits as well.

Reply:
I agree that would be better if it is not a problem that it makes parquet-mr not conform to the standard (of course we can update parquet-mr).

Reply:
With a wording to similar to what I used parquet-mr would conform to the specification, since padding with zeros would not be required, only recommended.

Reply:
Using the definitions of [RFC2119](https://www.ietf.org/rfc/rfc2119.txt):

 - MUST: This word, or the terms "REQUIRED" or "SHALL", mean that the
   definition is an absolute requirement of the specification.
- SHOULD: This word, or the adjective "RECOMMENDED", mean that there
   may exist valid reasons in particular circumstances to ignore a
   particular item, but the full implications must be understood and
   carefully weighed before choosing a different course.